test/_mystery-exit.js

```diff
@@ -0,0 +1,111 @@
+var workerCount = require('./helpers').workerCount;
+
+cluster.setupMaster({
+  exec: 'test/workers/null.js'
+});
+
+tap.test('from size 1 to 0', function(t) {
+  master.start({size: 1});
+  master.once('startWorker', function() {
+    master.setSize(0);
+  });
+
+  master.once('stopWorker', function(worker) {
+    t.assert(worker.process.pid, 'worker is valid');
+    worker.on('exit', function() {
+      t.equal(workerCount(), 0);
+      master.stop(t.end);
+    });
+  });
+});
+
+tap.test('from size 3 to 1', function(t) {
+  master.start({size: 3});
+  master.on('startWorker', function() {
+    if (workerCount() === 3) {
+      master.setSize(1);
+    }
+  });
+  master.once('stopWorker', function() {
+    t.equal(workerCount(), 2);
+    master.once('stopWorker', function() {
+      t.equal(workerCount(), 1);
+      master.stop(t.end);
+    });
+  });
+  console.log('trying to stay alive');
+
+  process.nextTick(function() {
+    console.log('next');
+  });
+  setImmediate(function() {
+    console.log('immed');
+  });
+  setTimeout(function() {
+    console.log('timeout');
+  }, 3000);
+  var a = require('net').createServer().listen(0, '127.0.0.1');
+  var b = require('net').createServer().listen(0, '::');
+  var c = require('net').createServer().listen(9876, '127.0.0.1');
+
+  console.log(a && b && c);
+
+  var dump = {
+    handles: function() {
+      var fmt = require('util').format;
+      var inspect = require('util').inspect;
+      var handles = process._getActiveHandles();
+
+      console.log('ActiveHandles:');
+      handles.forEach(function(h, i) {
+        console.log('%d: %s', i, pretty(h));
+      });
+
+      function pretty(h) {
+        if (h === process.stderr)
+          return fmt('process.stderr/type=%s/fd=%j', h._type, h.fd);
+        if (h === process.stdout)
+          return fmt('process.stdout/type=%s/fd=%j', h._type, h.fd);
+        if (h === process.stdin)
+          return fmt('process.stdin/type=%s/fd=%j', h._type, h.fd);
+        if ('_idleNext' in h)
+          return fmt('timer:msecs=%s:count=%d', h.msecs, timerCount(h));
+        if ('_connections' in h)
+          return fmt('server:%j', h.address());
+        return inspect(h, {depth: 1});
+      }
+
+      function timerCount(h) {
+        // XXX the circular list always has one extra entry for each setTimeout
+        // call, why?
+        var count = -1;
+        var first = h._idleNext;
+        var next = first;
+        while (next) {
+          next = next._idleNext;
+          count++;
+          if (next === first)
+            break;
+        }
+        return count;
+      }
+    },
+    requests: function() {
+      console.log('ActiveRequests:', process._getActiveRequests());
+    },
+  };
+  dump.handles();
+  dump.requests();
+  process.on('exit', function(code) {
+    console.log('exit: code %d', code);
+  });
+  process.on('beforeExit', function(code) {
+    console.log('exit: code %d', code);
+  });
+  console.log('bye?');
```

Comment:
yes, that's what is happening, thanks for the suggestion.

tap is trapping the exit event, and in the exit event, running the next test. this doesn't go so well.